crates/services/gas_price_service/src/lib.rs

```diff
@@ -152,21 +186,43 @@ where
                 tracing::debug!("Stopping gas price service");
                 should_continue = false;
             }
-            new_algo = self.update_algorithm.next() => {
-                let new_algo = new_algo?;
+            da_block_costs = self.da_block_costs_provider.recv() => {
+                self.da_block_costs = match da_block_costs {
+                    Ok(da_block_costs) => Some(da_block_costs),
+                    Err(err) => {
+                        tracing::error!("Failed to get da block costs: {:?}", err);
+                        None
+                    }
+                };
+                should_continue = true;
+            }
+            l2_block = self.l2_block_source.get_l2_block() => {
+                let l2_block = l2_block?;
+                let da_block_costs = self.da_block_costs.clone();
+                let next_algo = self.update_algorithm.next(l2_block, da_block_costs).await?;
                 tracing::debug!("Updating gas price algorithm");
-                self.update(new_algo).await;
+                self.update(next_algo).await;
                 should_continue = true;
             }
         }
         Ok(should_continue)
     }
 
     async fn shutdown(mut self) -> anyhow::Result<()> {
-        while let Some(new_algo) = self.update_algorithm.next().now_or_never() {
-            let new_algo = new_algo?;
-            tracing::debug!("Updating gas price algorithm");
-            self.update(new_algo).await;
+        self.da_block_costs_provider.stop_and_await().await?;
+
+        let l2_block = self.l2_block_source.get_l2_block().now_or_never();
+        if let Some(Ok(l2_block)) = l2_block {
+            let da_block_costs = self.da_block_costs.clone();
+            if let Some(new_algo) = self
+                .update_algorithm
+                .next(l2_block, da_block_costs)
+                .now_or_never()
+            {
+                let new_algo = new_algo?;
+                tracing::debug!("Updating gas price algorithm");
+                self.update(new_algo).await;
+            }
         }
```

Comment:
It should be a loop, since we can have more blocks than one. Plus it would be nice to avoid logic duplication, could you move the same logic into the function and reuse it here and in `run` function?

Reply:
avoided duplication in 911c06f71

i didn't understand why this should be a loop since we're shutting down here :)